### DISCUSSION TITLE
Shallow routing loses state, contrary to the docs

### DISCUSSION THREAD

**Discussion by flipperbw** [vercel/next.js#49221] [Help] [answered]
### Summary

I am trying to update the query parameters of a route (not dynamic, same page) without losing state information. There is a lot of contradictory information out there.

The docs (https://nextjs.org/docs/pages/building-your-application/routing/linking-and-navigating) say:

"You'll receive the updated pathname and the query via the [router object](https://nextjs.org/docs/pages/api-reference/functions/use-router#router-object) (added by [useRouter](https://nextjs.org/docs/pages/api-reference/functions/use-router#useRouter) or [withRouter](https://nextjs.org/docs/pages/api-reference/functions/use-router#withRouter)), **without losing state**....The URL will get updated to /?counter=10. and **the page won't get replaced**, only the state of the route is changed."

This appears to simply be untrue unless I am doing something wrong. The page reloads and I lose all state information.

A simple example, a file called `test.js` in the `pages` directory:

```test.js
import { useRouter } from 'next/router';
import { useState } from 'react';

export default function Test() {
    const [x, setX] = useState(1);
    const { push } = useRouter();

    return (
        <>
            <div>{x}</div>
            <button onClick={() => setX(2)}>set</button>
            <button onClick={() => push({ query: { y: 2 } }, undefined, { shallow: true })}>nav</button>
        </>
    );
}
```

Click the "set" button and the state is updated to "2". Now click "nav" and watch as the state gets reset to "1".

### Additional information

```js
Next info:

 Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 18.14.0
    Relevant packages:
      next: 13.3.4
      eslint-config-next: 13.3.4
      react: 18.2.0
      react-dom: 18.2.0
```
```


### Example

_No response_

**Comment by icyJoseph** [comment, marked as answer]
That's strange to say the least. Just tried it out on Stackblitz without issue: 

- https://stackblitz.com/edit/nextjs-kqes9w?file=pages%2Ftest.js,pages%2Findex.js

The funny part if that you don't even need to mark it as shallow. Shallow is only meant to prevent the server side functions from running.

State should be preserved. Do you think you could put together a https://stackblitz.com/ example of your issue?

I also tried with a fresh app using `create-next-app` no issue.

https://user-images.githubusercontent.com/21013447/236265734-dc39debc-4d65-4d08-9083-1472edd920cc.mov



Maybe you are doing something "strange" in `_app`?

**Reply by Johnrobmiller** [reply]
It appears that this might have been mistakenly marked as an answer. I could not find a solution to the problem in this response.

**Comment by flipperbw** [comment]
Sigh...yep. I sure was. Although I'm honestly not sure *why* exactly, here is what was happening in `_app.js`:

I had a wrapper component like this:

```
const ComponentWrapper = () => {
        const needsAuth = pathname.startsWith('/user/');
        return needsAuth ? (
            <AuthenticationGuard component={UserLayout}>
                <Component {...pageProps} />
            </AuthenticationGuard>
        ) : (
            <Component {...pageProps} />
        );
    };
```

So that the page would require a login (with a layout) if it was a `/user/` page. When I simply put this code directly in the return function (without a wrapper) everything works as expected.

**Reply by icyJoseph** [reply]
I'm out of office now, but maybe this? https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#component-types-and-reconciliation

Looks like that from your snippet. Anyhow, don't forget to mark as answered :) and ask further if you like.

**Reply by icyJoseph** [reply]
Just for completeness sake, yes it was that. You were recreating the component on every render, and that resets state.

- https://react.dev/learn/preserving-and-resetting-state

```jsx
import type { AppProps } from "next/app";

export default function App({ Component, pageProps }: AppProps) {
 // ❌❌❌ this is problematic and should always be avoided ❌❌❌
  const ComponentWrapper = () => {
    return <Component {...pageProps} />;
  };
  
  // the correct way
  return <Component {...pageProps} />;

  // A hacky way that also works
  // return ComponentWrapper();

  // this one creates a new instance on every render, which means state is recreated
  // it is essentially: Different components at the same position reset state 
  // return <ComponentWrapper />;
}
```

If you still want the component wrapper:

```jsx
import type { AppProps } from "next/app";

const ComponentWrapper = ({
  Component,
  pageProps,
}: Pick<AppProps, "Component" | "pageProps">) => {
  // do your business logic checks
  return <Component {...pageProps} />;
};
export default function App({ Component, pageProps }: AppProps) {
  return <ComponentWrapper Component={Component} pageProps={pageProps} />;
}

```

**Reply by Johnrobmiller** [reply]
I replicated this bug in my own project, but sadly, I found no places where component wrappers were being used like in the code examples above. Perhaps there is more than one cause of this problem.

**Reply by hedysnike** [reply]
I love next js BOTH pages and app router,

but app router without shallow routing is like holding a gun without bullets or

like super jet without any fuel

I assume vercel team hasn't added it because of maybe middleware or lack of localization support ?

**Comment by Johnrobmiller** [comment]
I found a stack overflow post that provides a solution that was very helpful to me to solve this issue, and maybe it will be helpful to other people as well (no guarentees)

https://stackoverflow.com/questions/64548544/how-to-change-route-without-re-rendering-in-next-js

EDIT: actually don't use that solution. It causes some weird behaviors when the Next.js router and the browser history go out of sync.
